src/librustc_mir/monomorphize/collector.rs

```diff
@@ -408,10 +406,12 @@ fn record_accesses<'tcx>(
         mono_item.instantiation_mode(tcx) == InstantiationMode::LocalCopy
     };
 
-    let accesses =
-        callees.into_iter().map(|mono_item| (*mono_item, is_inlining_candidate(mono_item)));
+    let accesses: SmallVec<[_; 128]> = callees
```

Comment:
We should add a comment here about the deadlock between `is_inlining_candidate` and `inlining_map`, if I'm following correctly.

Reply:
Yes, a deadlock definitely seems noteworthy.

Reply:
I'm not sure what to note here other than don't call random code in locks =P

Reply:
It is non-obvious to me at least when glancing at the code that there is a lock involved (and that is_inlining_candidate also takes it, if I'm recalling correctly). Explicitly noting that would be good.

Reply:
Everything pretty much uses a lock unless you know otherwise. We can't have explicit notes on all function calls.

Reply:
My main concern is that someone will drop the SmallVec here as an "optimization". I also feel like the argument about all functions potentially locking is a bit odd - if it were really true, then we'd be unable to review code well. I might even argue that a refactor which explicitly passes the lock in would be helpful here, if possible; it's hopefully rare that we have such unexpected locking.